File: examples/mnist.py
# 97% accuracy on MNIST test set after 200 epochs

import torch
import torch.nn as nn
from torch.utils.data import DataLoader
from torchvision import datasets, transforms

model = nn.Sequential(
    nn.Linear(784, 128),
    nn.ReLU(),
    nn.Linear(128, 10),
)

transform = transforms.Compose([transforms.ToTensor(), transforms.Normalize((0.1307,), (0.3081,))])
trainset = datasets.MNIST(root='./data', train=True, download=True, transform=transform)
testset = datasets.MNIST(root='./data', train=False, download=True, transform=transform)

# train without batch
train_loader = DataLoader(trainset, batch_size=1, shuffle=True)
test_loader = DataLoader(testset, batch_size=1, shuffle=True)

epochs = 200
criterion = nn.CrossEntropyLoss()
# Adam optimizer
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)

# train
for data, target in train_loader:
    # Flatten MNIST images into a 784 long vector
    data = data.view(data.shape[0], -1)
    # Training pass
    optimizer.zero_grad()
    output = model(data)
    loss = criterion(output, target)
    # This is where the model learns by backpropagating
    loss.backward()
    # And optimizes its weights here
    optimizer.step()
print("Training loss: {}".format(loss.item()))

# test
correct_count, all_count = 0, 0
for data, target in test_loader:
    data = data.view(data.shape[0], -1)
    output = model(data)
    _, pred = torch.max(output, 1)
    all_count += target.size(0)
    correct_count += (pred == target).sum()
print("Number Of Images Tested =", all_count)
print("Model Accuracy =", (correct_count / all_count).item())

File: examples/mnist.ts
import nn from "../torch/nn.ts";
import { load_mnist } from "../torch/vision.ts";

const path = new URL("./data", import.meta.url).pathname;

const [train_images, train_labels] = await load_mnist(path, true);
const [test_images, test_labels] = await load_mnist(path, false);

const input_size = 784
const hidden_sizes = [128, 64]
const output_size = 10

const optim = nn.optim.adam().build(1e-3);
const model = new nn.Sequential(
    nn.Linear(input_size, hidden_sizes[0], { bias: true, optim }),
    nn.ReLU(),
    nn.Linear(hidden_sizes[0], output_size, { bias: true, optim }),
);

for (let i = 0; i < 100; i++) {
    const loss = model.forward(train_images, optim).log_softmax(-1).nll_loss(train_labels);

    optim.zero_grad();
    loss.backward();
    optim.step();

    const test_accuracy = model.forward(test_images).argmax(1).eq(test_labels).to_float_type().mean();
    
    loss.print();
    test_accuracy.print();
}
